### PR TITLE
Correct primal of Matrix from UniformScaling & handle sum of boolean arrays

### PR DESCRIPTION
**Pull request by oxinabox**
Can't do AD on bools.

This was breaking code like:
```
using Distances, Zygote
function o(x)
           d = pairwise(Euclidean(), x, dims=2)
           return sum(iszero, d)
 end

gradient(o, rand(5,3))
````

since the sum was returning an `Array{Nothing}`.
Possibly we should post process those and normalize them into `nothing` more generally.
but to fix this particular case, just making sum of arrays of bools seems like a good start.

I would fix this in ChainRules but because Zygote defines `sum(::AbstractArray)` with a better definition than what ChainRules has right now, that will always take precidence.

### CODE REVIEW

File: src/lib/array.jl
Comment (molet) [inline src/lib/array.jl:233]: What is about other operations like `prod`?

Comment (oxinabox) [reply src/lib/array.jl:233]: Yes, this is not a general solution this just solves the `sum` case.

Comment (molet) [reply src/lib/array.jl:233]: I see, ok.

**Review by molet — APPROVED** [review]
LGTM.

**Comment by molet** [comment]
However, there are some tests failing now.

**Comment by oxinabox** [comment]
Wow the tests are failing for reasons that are not related to this PR,
except for the fact that this PR reveals them.

Zygote breaks the forward pass of `k*I`


```julia
julia> g(x) = Matrix(x*I, 2,2)
g (generic function with 1 method)

julia> g(2.5)
2×2 Array{Float64,2}:
 2.5  0.0
 0.0  2.5

julia> Zygote._pullback(g, 2.5)
(Bool[1 0; 0 1], ∂(g))
```

Which is a much more significant issue.

@mohamed82008 I will put fix in this PR if you could review

**Comment by oxinabox** [comment]
Tests are now only failing because tests fail in general in Julia 1.5

**Comment by DhairyaLGandhi** [comment]
Looks fine to me

**Comment by mcabbott** [comment]
This touches integer promotion rules. Is it obvious we want to treat `Bool` specially? 

The test you are adding `@test gradient(x->sum(x .== 0.0), [1.2, 0.2, 0.0, -1.1, 100.0]) == (nothing,)` is I think for #724, the problem isn't sum, it's the broadcast. #793 is a possible fix, which as you say processes `Array{Nothing}` into `nothing`.

**Comment by oxinabox** [comment]
> This touches integer promotion rules. Is it obvious we want to treat Bool specially?

I think so.
While it is not unusual to treat other integers as being actually embedded floats,
it is unusual to treat bools as.

> the problem isn't sum, it's the broadcast

It's both.
No matter where the array of Bools came from it should not had AD done to it.
But also things that output arrays of bools (like broadcast) should have be returning `nothing` to indicate not differentiable.

**Comment by mcabbott** [comment]
Have you got another example which produces an array of Bools, besides the broadcast? 

In #724's example, the array of bools never reaches `sum`, so it does have to be dealt with earlier. 

Re promotion, I'm not really taking a position, maybe they should be special. Just observing that it's a philosophical step, not a tiny bugfix. I don't think they are a special case elsewhere.

**Comment by oxinabox** [comment]
I would not say this is a tiny bug fix it is a feature enhancement.

But the other fix on the primal of Matrix from uniform scaling is a significant bug fix.
and I only found it because of the other change.

Which is itself an example of a time one might end up summing an array of bools, i guess.
Since our tests were. (though they were only doing so because of a bug this PR fixes)

```
julia> Matrix(I, 2,2)
2×2 Array{Bool,2}:
 1  0
 0  1

julia> sum(Matrix(I, 2,2))
2
```

**Comment by mcabbott** [comment]
Indeed, well found, and there could be more bugs of this class, without sum. Maybe disabling the gradient for things like `exp.(Matrix(I, 2,2))` and `map(>(0), rand(10))` might unearth more.

**Comment by oxinabox** [comment]
The real fix is to test things with ChainRulesTestUtils.jl
Which checks everything properly.

Anyway, I can split this PR if J have to.
Do I have to, or can this be merged?

**Comment by DhairyaLGandhi** [comment]
I think it's fine. We should also merge the updates needed for julia 1.5

**Comment by CarloLucibello** [comment]
> While it is not unusual to treat other integers as being actually embedded floats,
it is unusual to treat bools as.

I think this approach is the most reasonable one. 
I'm merging this as it is better to fix immediately this particular but likely common case

bors r+

**Comment by oxinabox** [comment]
@CarloLucibello 
Not a big deal but ideally you would have approved this PR and then left it to be to trigger Bors to merge it.
Since I have merge rights.

See https://github.com/SciML/ColPrac

 > PRs by people with merge rights should not be merged by people other than the author (just approved).

In case I had some last things left to do before merging.

Idk if I would have caught it in my own final check before merge, but the PR seems to have mistakenly included a submodule from the dev folder.

**Pull request merged**
